marketplacetests/marketplace/app.py

```diff
@@ -64,12 +64,30 @@ def offline_message_text(self):
         self.wait_for_element_displayed(*self._offline_message_locator)
         return self.marionette.find_element(*self._offline_message_locator).text
 
+    @property
+    def notification_message_text(self):
+        return self.marionette.execute_script('return window.document.getElementById("notification-content").innerHTML;')
+
+    def check_notification_message_for_text(self, text):
+        current_text = self.notification_message_text
+        # TODO: Added logging to help debug common intermittent failures
+        # This can be removed in the future if the intermittents disappear
+        print 'Looking for: "%s" in notification. Found: "%s"' % (text, current_text)
+        return text in current_text
+
     def wait_for_notification_message(self, message):
+        """This will wait for the specified message to appear in the DOM element
+           for the notification message, not for that message to be visible.
+
+           This is required as often the message is no longer visible when we check,
+           but the expected text still exists in the DOM element.
+
+           This is also the reason that we wait for the element to no longer be
+           visible at the end of this method, whereas we do not wait for it to
+           first be visible.
+        """
+        Wait(self.marionette).until(lambda m: self.check_notification_message_for_text(message))
```

Comment:
I prefer without the lambda if you need to pass more than just the marionette object, as I actually think it's cleaner. Basically you're defining your own expect object, which should be only needed in exceptional circumstances. If we start to notice patterns then we could either create a support class or extend the marionette driver. In this case it would be something like `expect.element_inner_html_contains(element, text)`. I actually prefer this, it's more generic and avoids repeatedly looking up the element in the wait. We can also stash it in another file of custom excepts:

```python
def wait_for_notification_message(self, message):
    notification = marionette.find_element(*self.app._notification_locator)
    Wait(self.marionette).until(self._element_inner_html_contains(notification, message))

class _element_inner_html_contains(object):

    def __init__(self, element, text):
        self.element = element
        self.needle = text

    def __call__(self, marionette):
        haystack = marionette.execute_script('return arguments[0].innerHTML;', [self.element])
        marionette.log('Looking for "%s" in "%s"' % (self.needle, haystack))
        return self.needle in haystack
```